src/main/java/com/faforever/client/game/GameService.java

```diff
@@ -312,6 +319,12 @@ public ReadOnlyBooleanProperty gameRunningProperty() {
       return completedFuture(null);
     }
 
+    if (!preferencesService.isGamePathValid()) {
+      CompletableFuture<Path> gameDirectoryFuture = new CompletableFuture<>();
+      eventBus.post(new GameDirectoryChooseEvent(gameDirectoryFuture));
+      return gameDirectoryFuture.thenCompose(path -> joinGame(game, password));
+    }
+
```

Comment:
This code is duplicated 6 times, I'd extract the first 2 lines to a method returning the gameDirectoryFuture.
```
if (!preferencesService.isGamePathValid()) {
    return ensureGameDirectorySet().thenAccept(p -> joinGame()/hostGame/whatever);
}
```

Reply:
true

Reply:
Acctually no it depends on the method look closely to the differences

Reply:
There is not much to refactor